### PR TITLE
Match module wildcards

### LINKED ISSUE #402: Allow module wildcards
In large projects you often want to enforce explicit silos of modules. These layers reduce the chance of circular dependencies and keep things a bit more sane. The module restriction feature could work perfectly for this if it had wildcards.

```
- modules:
  - {name: [Foo.Bar.*], within: [Foo.Bar.*, Foo]}
  - {name: [Foo.Qux.*], within: [Foo.Qux.*, Foo]}
```

This is likely a weaker form of the intention behind https://github.com/ndmitchell/lasagna

### PR DESCRIPTION
**Pull request by tfausak**
Partly addresses some of #402.

I had never looked at the HLint codebase before this morning, so I can't say if these are good changes or not. That being said this turned out to be a relatively easy change. And I added a test case for it.

### CODE REVIEW

File: src/Apply.hs [hunk 1]
```diff
@@ -113,5 +114,23 @@ classify xs i = let s = foldl' (f i) (ideaSeverity i) xs in s `seq` i{ideaSeveri
...
+-- True
+-- >>> isSpec "Namespaced.ExampleSpec"
+-- True
+-- >>> isSpec "Deeply.Nested.ExampleSpec"
+-- True
+--
+-- See this issue for details: <https://github.com/ndmitchell/hlint/issues/402>.
+wildcardMatch :: FilePattern -> String -> Bool
+wildcardMatch p m =
+    let f x = if x == '.' then '/' else x
```
Comment (ndmitchell) [inline src/Apply.hs:135]: `let f = replace "." "/"` would be simpler here and let you avoid the `fmap` below.

File: src/Apply.hs [hunk 2]
```diff
@@ -113,5 +114,23 @@ classify xs i = let s = foldl' (f i) (ideaSeverity i) xs in s `seq` i{ideaSeveri
...
+    in '*' `elem` p && fmap f p ?== fmap f m
```
Comment (ndmitchell) [inline src/Apply.hs:136]: Is the `*` test is a "cheap test" to check if it works? If yes, but has no functional difference, a comment would be good. If no, and it has a real semantic difference, what is it?

File: src/Config/Yaml.hs [hunk 3]
```diff
@@ -304,14 +304,18 @@ parseRestrict restrictType v = do
...
-        L _ (HsVar _ (L _ (Qual mod x))) -> pure $ f (moduleNameString mod) (occNameString x)
-        _ -> parseFail v "Bad classification rule"
-    where
-        f mod name@(c:_) | isUpper c = [(mod,name),(mod ++ ['.' | mod /= ""] ++ name, "")]
-        f mod name = [(mod, name)]
+    s <- parseString v
+    if '*' `elem` s
+        then pure [(s, "")]
+        else do
+            x <- parseGHC parseExpGhcWithMode v
```
Comment (ndmitchell) [inline src/Config/Yaml.hs:311]: Urk - this whole code is pretty grim. But it was before you got to it, so no complaint about this diff at all!

File: tests/wildcard.test [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---------------------------------------------------------------------
+RUN tests/wildcard.hs --hint=data/wildcard.yaml
+FILE tests/wildcard.hs
+module Spec where
+main = print $ (\ _ -> 'a') ()
+OUTPUT
+No hints
```
Comment (ndmitchell) [inline tests/wildcard.test:7]: I'd write this as two tests in the same file - one with `module SpecNo` that still triggers const, That makes sure you didn't just write `False` anywhere but are actually matching the wildcards.

**Review by ndmitchell — APPROVED** [review]
Amazing work. A few minor minor comments, but this looks good and a lot simpler than I expected.

File: src/Apply.hs (hunk 2 above)
Comment (tfausak) [reply src/Apply.hs:136]: Yes, it was meant as a cheap test to see if we should even bother trying a wildcard match. However now that I think about it we should be able to combine the regular and wildcard matchers. A regular matcher is just a wildcard that doesn't have any `*`s in it.

**Comment by tfausak** [comment]
I said that this fixes #402, but really it only addresses the `within` part of it. I forgot there was another part about, for example, module aliases:

``` yaml
- modules:
    - { name: 'Biz.Model.*', as: Model }
```

Perhaps another PR could tackle that part of the problem.

**Comment by ndmitchell** [comment]
Thanks, that's great!

**Pull request merged**